### PR TITLE
external checks: add new network connection check

### PR DESCRIPTION
**Pull request by bavarianbidi**
network-connection-check allows to test if a connection to a network target
could be done (via tcp or udp) and a failed connection is expected.

### CODE REVIEW

File: cmd/network-connection-check/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+The check runs every 2 minutes (spec.runInterval), with a check timeout set to 15 minutes (spec.timeout). If the check
+does not complete within the given timeout and target should be reachable it will report a timeout error on the status page. The check takes in a
+`CHECK_POD_TIMEOUT` environment variable that ensures the the pod runs the connection lookup within the timeout.
+If the check does not complete within the given timeout and target should not be reachable it will report a success on the status page.
+
+To verify connections, apply another KHCheck configuration file with a different `CONNECTION_TARGET` environment variable.
+`CONNECTION_TARGET` accepts `tcp://` or `udp://` (e.g. `udp://10.16.12.10:9000` or `tcp://10.16.12.11:8080`)
+
+#### DNS Status Check Kube Spec:
```
Comment (NissesSenap) [inline cmd/network-connection-check/README.md:22]: Should probably say network-connection-check

File: deploy/helm/kuberhealthy/values.yaml [hunk 2]
```diff
@@ -106,6 +106,16 @@ check:
       tag: v1.2.2
     extraEnvs:
     nodeSelector: {}
+  networkConnection:
```
Comment (integrii) [inline deploy/helm/kuberhealthy/values.yaml:109]: We don't have to add every new check to the helm chart here - we can simply add them to the check registry.

Actually, I don't even mind if the check is sourced from another git repository and image host entirely!  This means you could self-publish this check and then open a small PR to update the `EXTERNAL_CHECKS_REGISTRY.md` file.

File: deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+---
+apiVersion: comcast.github.io/v1
+kind: KuberhealthyCheck
+metadata:
+  name: network-connection-check
+  namespace: {{ .Release.Namespace }}
+spec:
+  runInterval:  {{ .Values.check.networkConnection.runInterval }}
+  timeout: {{ .Values.check.networkConnection.timeout }}
+  podSpec:
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml:11]:
```suggestion
  podSpec:
      {{- if .Values.securityContext.enabled }}
    securityContext:
      runAsUser: {{ .Values.securityContext.runAsUser }}
      fsGroup: {{ .Values.securityContext.fsGroup }}
    {{- end}}
```

File: deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      env:
+        - name: CHECK_TIME_LIMIT
+          value: {{ .Values.check.networkConnection.timeout }}
+{{- range $key, $value := .Values.check.networkConnection.extraEnvs }}
+        - name: {{ $key }}
+          value: {{ $value | quote }}
+{{- end }}
+      resources:
+        requests:
+          memory: 5Mi
```
Comment (integrii) [inline deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml:25]:
```suggestion
          memory: 5Mi
        {{- if .Values.securityContext.enabled }}
        securityContext:
          allowPrivilegeEscalation: {{ .Values.securityContext.allowPrivilegeEscalation }}
          readOnlyRootFilesystem: {{ .Values.securityContext.readOnlyRootFilesystem }}
        {{- end}}
```

File: cmd/network-connection-check/main.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+type Checker struct {
+	client            *kubernetes.Clientset
+	connectionTarget  string
+	targetUnreachable bool
+}
+
+func init() {
+
+	// Grab and verify environment variables and set them as global vars
+	CheckTimeout := os.Getenv("CHECK_POD_TIMEOUT")
```
Comment (jonnydawg) [inline cmd/network-connection-check/main.go:50]: I think checker pods are able to calculate how much time they have for a check run now with the `KH_CHECK_RUN_DEADLINE` environment variable which will give you a more accurate check deadline. This is still appropriate though.

Comment (integrii) [reply cmd/network-connection-check/main.go:50]: #450

Comment (bavarianbidi) [reply cmd/network-connection-check/main.go:50]: wasn't aware of this ENV ...
... have to rework this PR.
I would update the documentation regarding all injected ENV vars in this PR.

**Comment by jonnydawg** [comment]
Some deadline documentation was touched in #452 and #454

**Comment by integrii** [comment]
This one is currently waiting for the feedback about the check pod timeout ENV var to be integrated before merging.

File: deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml (hunk 3 above)
Comment (bavarianbidi) [reply deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml:11]: from my point of view we don't need the `securityContext` here.
* the application itself doesn't need any extended permissions (only does a tcp/udp connection check, no additional file operations).
* container is build with a numeric user and point to `nobody` (https://github.com/Comcast/kuberhealthy/pull/443/files#diff-b80ba549f0d3f00bfc542d54b0e03f7cR10) 
* if a `psp` is deployed where the usage of `root` is forbidden, kubernetes can schedule pods without `securityContext` when the `USER` field is a numeric value.

File: deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml (hunk 4 above)
Comment (bavarianbidi) [reply deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml:25]: from my point of view we don't need the `securityContext` here.
* the application itself doesn't need any extended permissions (only does a tcp/udp connection check, no additional file operations).
* container is build with a numeric user and point to `nobody` (https://github.com/Comcast/kuberhealthy/pull/443/files#diff-b80ba549f0d3f00bfc542d54b0e03f7cR10) 
* if a `psp` is deployed where the usage of `root` is forbidden, kubernetes can schedule pods without `securityContext` when the `USER` field is a numeric value.

**Comment by bavarianbidi** [comment]
> This one is currently waiting for the feedback about the check pod timeout ENV var to be integrated before merging.

was AFK the last days, will test tomorrow morning and give feedback here

**Comment by bavarianbidi** [comment]
Will Switch back to the previous implementation with a separate `TIMEOUT` variable...
... the current version sometimes fail if connection target is expected to be unreachable. time until the check pod will start does take a while ( image has to be pulled, ... ) and the current hard coded `-5s` isn't flexible for this and a separate parameter make sense (IMHO)

**Comment by bavarianbidi** [comment]
@jonnydawg / @integrii 

could you please take a look again ;-)

File: deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml (hunk 3 above)
Comment (integrii) [reply deploy/helm/kuberhealthy/templates/khcheck-network-connection.yaml:11]: Gotcha.  As long as it works in these kind of locked down environments, I don't have a strong opinion.

**Pull request merged**